prboom2/src/gl_main.c

```diff
@@ -3232,41 +3232,59 @@ void gld_DrawScene(player_t *player)
 
   if (gld_drawinfo.num_items[GLDIT_TWALL] > 0 || gld_drawinfo.num_items[GLDIT_TSPRITE] > 0)
   {
-    if (gld_drawinfo.num_items[GLDIT_TWALL] > 0)
-    {
-      // if translucency percentage is less than 50,
-      // then all translucent textures and sprites disappear completely
-      // without this line
-      glAlphaFunc(GL_GREATER, 0.0f);
+      int twall_idx   = gld_drawinfo.num_items[GLDIT_TWALL] - 1;
+      int tsprite_idx = gld_drawinfo.num_items[GLDIT_TSPRITE] - 1;
+
+      if (tsprite_idx > 0)
+          gld_DrawItemsSortSprites(GLDIT_TSPRITE);
 
-      // transparent walls
-      for (i = gld_drawinfo.num_items[GLDIT_TWALL] - 1; i >= 0; i--)
+      while (twall_idx >= 0 || tsprite_idx >= 0 )
       {
-        gld_SetFog(gld_drawinfo.items[GLDIT_TWALL][i].item.wall->fogdensity);
-        gld_ProcessWall(gld_drawinfo.items[GLDIT_TWALL][i].item.wall);
-      }
-    }
+          dboolean draw_tsprite = true;
 
-    glEnable(GL_ALPHA_TEST);
+          /* find out which is next to draw */
+          if (twall_idx >= 0 && tsprite_idx >= 0)
+          {
+              /* both are left to draw, determine
+               * which is farther */
+              seg_t *twseg = gld_drawinfo.items[GLDIT_TWALL][twall_idx].item.wall->seg;
+              /* reconstruct the sprite xy */
+              fixed_t tsx = gld_drawinfo.items[GLDIT_TSPRITE][tsprite_idx].item.sprite->xy & 0xFFFF0000;
+              fixed_t tsy = (gld_drawinfo.items[GLDIT_TSPRITE][tsprite_idx].item.sprite->xy & 0x0000FFFF) << 16;
+
+              if (R_PointOnSegSide(tsx, tsy, twseg))
+              {
```

Comment:
Is it true that this rather expensive calculation is only ever executed if there are both translucent walls and sprites in the rendered scene? And, now comes the obvious question, is it somehow possible to cache this information from an earlier stage of the rendering process? I mean, the software renderer also knows if the sprite is in front of the wall or behind it, but this is probably unrelated.

Reply:
Correct - this code is only executed if both translucent walls and sprites (things) are present in the scene at the same time.

To cache the relative distances would require depth sorting of both object types based on a common reference point. I think it would require essentially the same calculations. I am not sure it would save too much in the end.

Reply:
> Yes, it seems to be working, awesome! However, it might be due to my eyes, though, I find it hard to tell which sphere is in front of the wall and which is behind. To me it looks like they are all drawn as if they were behind the wall.

Ah, I think you are right - the logic is reversed in that case.

Oh there is one limitation I should mention -- the overlap is determined based on the center of the thing and not the true sprite edges. So when a rendered sprite is partially inside a translucent wall, it may not render as partially in and out of the wall at the same time and there is a "slicing" effect. This is perhaps something I can put as a future fix option.